File: crates/store/re_sdk_types/src/archetypes/ellipses2d_ext.rs
use super::Ellipses2D;
use crate::components::{HalfSize2D, Position2D};

impl Ellipses2D {
    /// Creates new [`Ellipses2D`] for circles with the given radii, centered around the local origin.
    // Note: This is not a `Radius` component because the `Radius` component is for
    // the on-screen sizes of lines and points.
    #[inline]
    #[doc(alias = "circle")]
    pub fn from_radii(radii: impl IntoIterator<Item = f32>) -> Self {
        Self::new(radii.into_iter().map(HalfSize2D::splat))
    }

    /// Creates new [`Ellipses2D`] for circles with [`Self::centers`] and the given radii.
    // Note: This is not a `Radius` component because the `Radius` component is for
    // the on-screen sizes of lines and points.
    #[inline]
    #[doc(alias = "circle")]
    pub fn from_centers_and_radii(
        centers: impl IntoIterator<Item = impl Into<Position2D>>,
        radii: impl IntoIterator<Item = f32>,
    ) -> Self {
        Self::new(radii.into_iter().map(HalfSize2D::splat)).with_centers(centers)
    }

    /// Creates new [`Ellipses2D`] with [`Self::half_sizes`] centered around the local origin.
    #[inline]
    pub fn from_half_sizes(half_sizes: impl IntoIterator<Item = impl Into<HalfSize2D>>) -> Self {
        Self::new(half_sizes)
    }

    /// Creates new [`Ellipses2D`] with [`Self::centers`] and [`Self::half_sizes`].
    #[inline]
    pub fn from_centers_and_half_sizes(
        centers: impl IntoIterator<Item = impl Into<Position2D>>,
        half_sizes: impl IntoIterator<Item = impl Into<HalfSize2D>>,
    ) -> Self {
        Self::new(half_sizes).with_centers(centers)
    }
}

File: rerun_cpp/src/rerun/archetypes/ellipses2d_ext.cpp
#include "ellipses2d.hpp"

#include "../collection_adapter_builtins.hpp"

// #define EDIT_EXTENSION

namespace rerun {
    namespace archetypes {

#ifdef EDIT_EXTENSION
        // <CODEGEN_COPY_TO_HEADER>

        /// Creates new `Ellipses2D` with `half_sizes` centered around the local origin.
        static Ellipses2D from_half_sizes(Collection<components::HalfSize2D> half_sizes) {
            return Ellipses2D().with_half_sizes(std::move(half_sizes));
        }

        /// Creates new `Ellipses2D` with `centers` and `half_sizes`.
        static Ellipses2D from_centers_and_half_sizes(
            Collection<components::Position2D> centers,
            Collection<components::HalfSize2D> half_sizes
        ) {
            return Ellipses2D()
                .with_half_sizes(std::move(half_sizes))
                .with_centers(std::move(centers));
        }

        // </CODEGEN_COPY_TO_HEADER>
#endif
    } // namespace archetypes
} // namespace rerun
